proposals/downloads_get_file_hash.md

```diff
@@ -0,0 +1,152 @@
+    callback GetFileHashCallback = void(DOMString hash);
+    [supportsPromises, permissions="downloads.getFileHash"] static void getFileHash(
+        long downloadId,
+        [legalValues=("SHA-256")] DOMString algorithm,
+        GetFileHashCallback callback);
+}
+```
+
+``` idl
+namespace downloads {
+  enum FILE_HASH_ALGO {
+    "SHA-256"
+  }
+
+  getFileHash: (downloadId: number, algorithm?: downloads.FILE_HASH_ALGO) => Promise<string>;
+}
+```
+
+### New Permissions
+
+|      Permission Added     | Suggested Warning |
+| ------------------------- | ----------------- |
+| `"downloads.getFileHash"` | `"View cryptographic digests of downloads"` |
+
+The new `"downloads.getFileHash"` permission is optional and is honored only if
+extension has `"downloads"` permission as well.
+
+### Manifest File Changes
+
+No thanges besides the new optional permission `"downloads.getFileHash"`.
+
+## Security and Privacy
+
+### Exposed Sensitive Data
+
+Hashes (cryptographic) are considered one-way operations, but they can easily
+be reversed to the original input when the number of possible inputs is
+sufficiently small. So extension may be able to infer the raw contents of the
+file if the possible space of file contents is small.
+A malicious extension could:
+ - Observe hashes of downloaded files and send them to a home server
+
+### Abuse Mitigations
+
+The API method will require an explicit permission with a permission warning.
+Browsers may apply the same mitigations they use for `DownloadItem.url` and
+`DownloadItem.finalUrl`.
+
+### Additional Security Considerations
+
+User Agents must not provide obsolete hash algorithms (like MD5 and SHA-1).
+User Agent must ever expose only hashes of complete files and never provide
+partial hash results while the file is being downloaded. This is necessary to
+prevent malicious or compromised extension from calculating incremental hashes
+and constructing a "hash oracle" to infer file contents one byte (or just a few
+bytes) at a time.
+
+## Alternatives
+
+### Existing Workarounds
+
+As of writing, extensions can calculte digest of a download by implementing
+a native host component which would read the desired file from disc directly.
+This approach is a significant elevation of priviledges (requires unsandbxed
+application), largely expands the complexity of implementing such extension
+(since the extension has to listen for download completion, notify the native
+componen, and ensure that the file does not change or move before the hash is
+calculated) and raises friction for users (due to complexity of installation).
+Additionally, calculating file content hash separately is slightly less
+efficient since it requires reading it back into memory.
+
+### Open Web API
+
+There are no equivalent APIs on the Open Web. Open Web has the concept of
+subresource integrity, but it applies only to page resources like images and
+other media and does not apply to downloaded files.
+
+## Implementation Notes
```

Comment:
@Rob--W One more question: should the encoded hash be upper-case or lower-case? I have no particular preference, just would like the spec to be unambiguous.

Reply:
Lowercase seems more common.

Reply:
I added point 4 "User Agents must encode (binary) hash digest into a hex string, using lower-case letters." Is this OK?